src/hotspot/share/classfile/bytecodeAssembler.hpp

```diff
@@ -75,11 +74,11 @@ class BytecodeCPEntry {
       u2 class_index;
       u2 name_and_type_index;
     } methodref;
-    uintptr_t hash;
+    uintptr_t value;
```

Comment:
Now you have drawn attention to this by renaming it, what exactly is it? The rest of the union I can understand as they are different kinds of CP entries but what is a "value"??

Reply:
Value is essentially (uintptr_t)utf8, or (uintptr_t)u2klass|u4name_and_type, or (uintptr_t)u2klass|u4methodref;  ie the value of the fields that are in the union.

Reply:
That is technically undefined behaviour.

Reply:
I believe that.  If this is also undefined behavior

  static unsigned hash(BytecodeCPEntry const& e0) {
    return (unsigned)(e0._tag ^ e0._u.value);
  }

Then this BytecodeCPEntry class needs to be restructured a bit, which shouldn't be in this PR.

Reply:
Not a review, just a drive-by comment.

I'm assuming the UB @dholmes-ora  is referring to is storing a value in one member of
a union and then reading from another, aka the "union trick". That is indeed
UB according to the C++ standard. It used to also be UB in C, but that was
changed in C99.

However, "Many compilers implement, as a non-standard language extension, the
ability to read inactive members of a union."
https://en.cppreference.com/w/cpp/language/union

And indeed, gcc and msvc both explicitly document providing that extension.  I
expect clang does too (for compatibility reasons), though I've not found the
words that say so.

The "union trick" is used in several places in HotSpot, sometimes even with
commentary about the UB per-the-standard but compiler supported.  Perhaps the
style guide should explicitly mention it, so I can avoid needing to repeat
this comment.

Reply:
https://bugs.openjdk.org/browse/JDK-8329818 - HotSpot Style Guide should explicitly permit the "union trick"

Reply:
Thanks @kimbarrett . I didn't recognise this code as "the union trick" as such, but it is - we use the `value` member to access the bit pattern of the real (primitive) value for hashing purposes. Certainly should be safe/harmless for the types involved.

Reply:
It's used for hashing and for equals comparison.  I wouldn't want to change one without the other.

Reply:
Maybe renaming it from 'hash' to 'value' wasn't a good idea.  The equals method makes sense in that it compares the 'hash' also.